src/parser/lex.rs

```diff
@@ -12,71 +12,1019 @@
+    static ref ALL_DURATION_BUT_YEAR_UNITS: HashSet<char> =
+        HashSet::from(['s', 'm', 'h', 'd', 'w']);
+    static ref SPACE_SET: HashSet<char> = HashSet::from([' ', '\t', '\n', '\r']);
+    static ref HEX_CHAR_SET: HashSet<char> = HashSet::from(['x', 'X']);
+    static ref SCI_CHAR_SET: HashSet<char> = HashSet::from(['e', 'E']);
+    static ref SIGN_CHAR_SET: HashSet<char> = HashSet::from(['+', '-']);
+    static ref NORMAL_ESCAPE_SYMBOL_SET: HashSet<char> = "abfnrtv\\".chars().into_iter().collect();
+    static ref STRING_SYMBOL_SET: HashSet<char> = HashSet::from(['"', '`', '\'']);
+}
 
 pub type LexemeType = DefaultLexeme<TokenType>;
-pub type Lexer<'a> = LRNonStreamingLexer<'a, 'a, LexemeType, TokenType>;
-
-// FIXME: this is just a demo Lexer, constructed for example
-pub fn lexer(s: &str) -> Lexer {
-    let mut start = 0;
-    let mut len = "node_cpu_seconds_total".len();
-    let metric_identifier_lexeme =
-        DefaultLexeme::new(crate::parser::T_METRIC_IDENTIFIER, start, len);
-
-    start += len;
-    len = "{".len();
-    let left_brace_lexeme = DefaultLexeme::new(crate::parser::T_LEFT_BRACE, start, len);
-
-    start += len;
-    len = "cpu".len();
-    let identifier1_lexeme = DefaultLexeme::new(crate::parser::T_IDENTIFIER, start, len);
-
-    start += len;
-    len = "=".len();
-    let eql1_lexeme = DefaultLexeme::new(crate::parser::T_EQL, start, len);
-
-    start += len;
-    len = "0".len();
-    let val1_lexeme = DefaultLexeme::new(crate::parser::T_STRING, start, len);
-
-    start += len;
-    len = ",".len();
-    let comma_lexeme = DefaultLexeme::new(crate::parser::T_COMMA, start, len);
-
-    start += len;
-    len = "mode".len();
-    let identifier2_lexeme = DefaultLexeme::new(crate::parser::T_IDENTIFIER, start, len);
-
-    start += len;
-    len = "=".len();
-    let eql2_lexeme = DefaultLexeme::new(crate::parser::T_EQL, start, len);
-
-    start += len;
-    len = "idel".len();
-    let val2_lexeme = DefaultLexeme::new(crate::parser::T_STRING, start, len);
-
-    start += len;
-    len = "}".len();
-    let right_brace_lexeme = DefaultLexeme::new(crate::parser::T_RIGHT_BRACE, start, len);
-
-    let lexemes = vec![
-        metric_identifier_lexeme,
-        left_brace_lexeme,
-        identifier1_lexeme,
-        eql1_lexeme,
-        val1_lexeme,
-        comma_lexeme,
-        identifier2_lexeme,
-        eql2_lexeme,
-        val2_lexeme,
-        right_brace_lexeme,
-    ]
-    .into_iter()
-    .map(Ok)
-    .collect();
-
-    LRNonStreamingLexer::new(s, lexemes, Vec::new())
+
+pub fn lexer(s: &str) -> Result<LRNonStreamingLexer<LexemeType, TokenType>, String> {
+    let lexemes: Vec<Result<LexemeType, String>> = Lexer::new(s).into_iter().collect();
+    match lexemes.last() {
+        Some(Err(info)) => Err(info.into()),
+        None => Err(format!("generated empty lexemes for {}", s)),
+        _ => {
```

Comment:
[https://pingcap.github.io/style-guide/rust/exprs.html](https://pingcap.github.io/style-guide/rust/exprs.html)
> Prefer to list patterns explicitly rather than using a wildcard (e.g., _) pattern. Rationale: fail fast if an enum changes.

```suggestion
        Some(Ok(_)) => {
```